src/Features/CSharp/Portable/SignatureHelp/LightweightOverloadResolution.cs

```diff
@@ -0,0 +1,254 @@
+        var argumentIndexToSave = GetArgumentIndex();
+        var foundParameterIndex = -1;
+        if (argumentIndexToSave >= 0)
+        {
+            foundParameterIndex = argumentToParameterMap[argumentIndexToSave];
+            if (foundParameterIndex < 0)
+                foundParameterIndex = FirstUnspecifiedParameter(ref argumentToParameterMap.AsRef());
+        }
+
+        Debug.Assert(foundParameterIndex < parameters.Length);
+
+        return (method, foundParameterIndex);
+    }
+
+    /// <summary>
+    /// Determines if the given argument is compatible with the given parameter
+    /// </summary>
+    private bool IsCompatibleArgument(ArgumentSyntax argument, IParameterSymbol parameter)
+    {
+        var parameterRefKind = parameter.RefKind;
+        if (parameterRefKind == RefKind.None)
+        {
+            if (IsEmptyArgument(argument.Expression))
+            {
+                // An argument left empty is considered to match any parameter
+                // M(1, $$)
+                // M(1, , 2$$)
+                return true;
+            }
+
+            var type = parameter.Type;
+            if (parameter.IsParams
+                && type is IArrayTypeSymbol arrayType
+                && semanticModel.ClassifyConversion(argument.Expression, arrayType.ElementType).IsImplicit)
+            {
+                return true;
+            }
+
+            return semanticModel.ClassifyConversion(argument.Expression, type).IsImplicit;
+        }
+
+        var argumentRefKind = argument.GetRefKind();
+        if (parameterRefKind == argumentRefKind)
+            return true;
+
+        // A by-value argument matches an `in` parameter
+        if (parameterRefKind == RefKind.In && argumentRefKind == RefKind.None)
+            return true;
+
+        return false;
+    }
+
+    // If the cursor is pointing at an argument for which we did not find the corresponding
+    // parameter, we will highlight the first unspecified parameter.
+    private int FirstUnspecifiedParameter(ref TemporaryArray<int> argumentToParameterMap)
+    {
+        using var specified = TemporaryArray<bool>.Empty;
+        for (var i = 0; i < arguments.Count; i++)
+            specified.Add(false);
+
+        for (var i = 0; i < arguments.Count; i++)
+        {
+            var parameterIndex = argumentToParameterMap[i];
+            if (parameterIndex >= 0 && parameterIndex < arguments.Count)
+                specified.AsRef()[parameterIndex] = true;
+        }
+
+        for (var i = 0; i < specified.Count; i++)
+        {
+            if (!specified[i])
+                return i;
+        }
+
+        return 0;
+    }
+
+    /// <summary>
+    /// Find the parameter index corresponding to each argument provided
+    /// </summary>
+    private bool TryPrepareArgumentToParameterMap(IMethodSymbol method, ref TemporaryArray<int> argumentToParameterMap)
+    {
+        Contract.ThrowIfTrue(argumentToParameterMap.Count != arguments.Count);
```

Comment:
Instead of passing in the array as a ref, could it be initialized here and returned as an out parameter?

Reply:
not while using it in a `using` scope properly.